lib/index.js

```diff
@@ -210,16 +246,48 @@ export function configureUnifiedLanguageServer(
     }
   }
 
-  connection.onInitialize(() => ({
-    capabilities: {
-      textDocumentSync: TextDocumentSyncKind.Full,
-      documentFormattingProvider: true,
-      codeActionProvider: {
-        codeActionKinds: [CodeActionKind.QuickFix],
-        resolveProvider: true
+  connection.onInitialize((event) => {
+    if (event.workspaceFolders) {
+      workspaces = event.workspaceFolders.map((d) => d.uri)
+    }
+
+    if ((!workspaces || workspaces.length === 0) && event.rootUri) {
+      workspaces = [event.rootUri]
+    }
+
+    if (
+      event.capabilities.workspace &&
+      event.capabilities.workspace.workspaceFolders
+    ) {
+      connection.workspace.onDidChangeWorkspaceFolders(function (event) {
+        const added = event.added.map((d) => d.uri)
+        const removed = new Set(event.removed.map((d) => d.uri))
+        // The `else` case should not occur: presumably, when the client says it
+        // supports workspace folders and sends notifications, it passed
+        // workspaces to `onInitialize` too.
+        /* c8 ignore next */
+        const current = workspaces || []
+
+        workspaces = [...current.filter((d) => !removed.has(d)), ...added]
```

Comment:
Given suggestions above, this would become:

```js
removeWorkspaces(event.removed.map((d) => d.uri))
addWorkspaces(event.added.map((d) => d.uri))
```